patches/api/0449-Add-Lifecycle-Event-system.patch

```diff
@@ -0,0 +1,577 @@
++public interface LifecycleEvent {
++}
+diff --git a/src/main/java/io/papermc/paper/plugin/lifecycle/event/types/LifecycleEventType.java b/src/main/java/io/papermc/paper/plugin/lifecycle/event/types/LifecycleEventType.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..b56a90705e6029034c576c1effcc5a85c740ee2b
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/plugin/lifecycle/event/types/LifecycleEventType.java
+@@ -0,0 +1,72 @@
++package io.papermc.paper.plugin.lifecycle.event.types;
++
++import io.papermc.paper.plugin.lifecycle.event.LifecycleEventManager;
++import io.papermc.paper.plugin.lifecycle.event.LifecycleEventOwner;
++import io.papermc.paper.plugin.lifecycle.event.handler.LifecycleEventHandler;
++import io.papermc.paper.plugin.lifecycle.event.handler.configuration.LifecycleEventHandlerConfiguration;
++import io.papermc.paper.plugin.lifecycle.event.handler.configuration.MonitorLifecycleEventHandlerConfiguration;
++import io.papermc.paper.plugin.lifecycle.event.handler.configuration.PrioritizedLifecycleEventHandlerConfiguration;
++import org.jetbrains.annotations.ApiStatus;
++import org.jetbrains.annotations.Contract;
++import org.jetbrains.annotations.NotNull;
++
++/**
++ * Base type for all types of lifecycle events. Differs from
++ * {@link LifecycleEvent} which is the actual event object whereas
++ * this is an object representing the type of the event. Used
++ * to construct subtypes of {@link LifecycleEventHandlerConfiguration} for
++ * use in {@link LifecycleEventManager}
++ *
++ * @param <O> the required owner type
++ * @param <E> the event object type
++ * @param <C> the configuration type
++ */
++@ApiStatus.Experimental
++@ApiStatus.NonExtendable
++public interface LifecycleEventType<O extends LifecycleEventOwner, E extends LifecycleEvent, C extends LifecycleEventHandlerConfiguration<O>> {
++
++    /**
++     * Gets the name of the lifecycle event.
++     *
++     * @return the name
++     */
++    @Contract(pure = true)
++    @NotNull String name();
++
++    /**
++     * Create a configuration for this event with the specified
++     * handler.
++     *
++     * @param handler the event handler
++     * @return a new configuration
++     * @see LifecycleEventManager#registerEventHandler(LifecycleEventHandlerConfiguration)
++     */
++    @Contract("_ -> new")
++    @NotNull C newHandler(@NotNull LifecycleEventHandler<? super E> handler);
++
++    /**
++     * Lifecycle event type that supports separate registration
++     * of handlers as "monitors" which are run last. Useful
++     * if a plugin wants to only observe the changes other handlers
++     * made.
++     *
++     * @param <O> the required owner type
++     * @param <E> the event object type
++     */
++    @ApiStatus.Experimental
++    @ApiStatus.NonExtendable
++    interface Monitor<O extends LifecycleEventOwner, E extends LifecycleEvent> extends LifecycleEventType<O, E, MonitorLifecycleEventHandlerConfiguration<O>> {
```

Comment:
`Monitor` as a name feels wrong here, it's more of a Monitorable event, though that's a mouthfull. Same for `Prioritized` below

Reply:
So you think Monitorable and Prioritizable?

Reply:
Yeah, not sure if there's better alternatives

Reply:
done